### PR TITLE
Fixing tar header and content order mismatch

### PR DESCRIPTION
**Pull request by morris-jason**
Experienced a situation where tar headers were invalid but content was still being written to the tar archive. Consuming the guest download in a tar reader was causing EOFs in the middle of the tar stream.

### CODE REVIEW

File: vim25/soap/client.go [hunk 1]
```diff
@@ -686,10 +686,7 @@ func (c *Client) Download(u *url.URL, param *Download) (io.ReadCloser, int64, er
 	}
 
 	r := res.Body
-
-	if strings.HasSuffix(u.Path, "/") {
```
Comment (dougm) [inline vim25/soap/client.go]: This was intentional, to avoid the slim but possible case of hitting the gzip header sequence when downloading a regular file.  Is there an issue with having the client use a trailing "/" to specify a directory?  

Is this what you meant by "missing backslashes"?

File: toolbox/hgfs/archive.go [hunk 2]
```diff
@@ -229,42 +229,37 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
...
-		_ = tw.WriteHeader(header)
+		if err := tw.WriteHeader(header); err != nil {
+			return err
+		}
 
-		if f != nil {
-			_, err = io.Copy(tw, f)
-			_ = f.Close()
+		content, err := os.Open(file)
+		defer content.Close()
```
Comment (dougm) [inline toolbox/hgfs/archive.go]: This will panic if os.Open returns an error.
And why the change to use defer?

File: toolbox/hgfs/archive.go [hunk 3]
```diff
@@ -229,42 +229,37 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
...
-		if header.Typeflag == tar.TypeReg && fi.Size() != 0 {
-			f, err = os.Open(file)
-			if err != nil {
-				if os.IsPermission(err) {
-					return nil
-				}
-				return err
-			}
+		if header.Typeflag != tar.TypeReg {
+			return nil
```
Comment (dougm) [inline toolbox/hgfs/archive.go]: Then we don't get tar entries for symlink or directories.

File: toolbox/hgfs/archive.go [hunk 4]
```diff
@@ -229,42 +229,37 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
...
 		}
 
-		_ = tw.WriteHeader(header)
+		if err := tw.WriteHeader(header); err != nil {
+			return err
```
Comment (dougm) [inline toolbox/hgfs/archive.go]: Was there a case where WriteHeader returned an error related to the header itself?  Otherwise, if the pipe was closed, then the io.Copy below would return the same error.

File: toolbox/hgfs/archive.go [hunk 5]
```diff
@@ -229,42 +229,37 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
...
 
-		if f != nil {
-			_, err = io.Copy(tw, f)
-			_ = f.Close()
+		content, err := os.Open(file)
+		defer content.Close()
+		if err != nil {
+			return err
+		}
+		if _, err := io.Copy(tw, content); err != nil {
```
Comment (dougm) [inline toolbox/hgfs/archive.go]: Why add the if statement here, rather than just return error as the original code did?

File: toolbox/hgfs/archive.go [hunk 6]
```diff
@@ -229,42 +229,37 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
 
 // archiveWrite writes the contents of the given source directory to the given tar.Writer.
 func archiveWrite(u *url.URL, tw *tar.Writer) error {
-	dir := filepath.Dir(u.Path)
 
 	return filepath.Walk(u.Path, func(file string, fi os.FileInfo, err error) error {
 		if err != nil {
-			return filepath.SkipDir
+			return err
```
Comment (dougm) [inline toolbox/hgfs/archive.go]: This was intentional, behaves like the tar command (tho tar prints a warning in this case).

File: toolbox/hgfs/archive.go (hunk 3 above)
Comment (morris-jason) [inline toolbox/hgfs/archive.go]: Yeah that was supposed to go after the write. Fixed.

File: toolbox/hgfs/archive.go (hunk 6 above)
Comment (morris-jason) [inline toolbox/hgfs/archive.go]: reverted.

File: vim25/soap/client.go (hunk 1 above)
Comment (morris-jason) [inline vim25/soap/client.go]: I'll add it back in and just guarantee the user enters a backslash on my side.

File: toolbox/hgfs/archive.go (hunk 2 above)
Comment (morris-jason) [inline toolbox/hgfs/archive.go]: Didn't know it would panic. reverting.

**Comment by morris-jason** [comment]
Updated the pr to be more accurate. Tests now passing.

File: toolbox/hgfs/archive.go (hunk 4 above)
Comment (morris-jason) [inline toolbox/hgfs/archive.go]: I was having issues on the consuming side with tar readers receiving EOFs for tar headers. I suspect there was an error writing the header, but the copy below still succeeded. That's why I'm returning here if there's an error.

File: toolbox/hgfs/archive.go [hunk 7]
```diff
@@ -229,42 +229,36 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
...
+			string(filepath.Separator))
 
-		var f *os.File
+		if err := tw.WriteHeader(header); err != nil {
+			return err
+		}
 
 		if header.Typeflag == tar.TypeReg && fi.Size() != 0 {
-			f, err = os.Open(file)
+			f, err := os.Open(file)
```
Comment (dougm) [inline toolbox/hgfs/archive.go]: Now the file isn't closed at all, so this leaks file descriptors.

The ignored errors from FileInfoHeader and WriteHeader were intentional, the additional checks aren't needed afaict.

And didn't see a response to this question in the first review "Can't tell how this is different without trying it, or what exactly was broken. Can you give some examples?"

Comment (morris-jason) [inline toolbox/hgfs/archive.go]: I was having issues on the consuming side with tar readers receiving EOFs for tar headers. I suspect there was an error writing the header, but the copy still still succeeded.

File: toolbox/hgfs/archive.go (hunk 4 above)
Comment (dougm) [inline toolbox/hgfs/archive.go]: You would get an "unexpected EOF" if InitializeTransferFromGuest was called without the trailing "/".  Otherwise, I'm not seeing why these changes are needed.  Just want to understand what the issue(s) are.

**Comment by morris-jason** [comment]
Closing, reproduced the errors regardless of these changes.
